### ISSUE TITLE
npm@3 given the same module list, not linking in same order

### ISSUE THREAD

**Issue by phated** [npm/npm#8995]
As per @iarna at https://github.com/gulpjs/gulp-cli/issues/7#issuecomment-121037259

I haven't tested this, but a gulp core contributor tested and linked results at https://github.com/gulpjs/gulp-cli/issues/7#issuecomment-120907981

**Comment by othiym23** [comment]
Thanks for the report, @phated. We're probably going to have to take a few passes at getting this resolved in a way that feels good to everybody, but I imagine @iarna has a solution in mind for this (and the other issue you filed related to this) in a nearish term.

**Comment by iarna** [comment]
I haven't been able to reproduce this one, but as an ordering bug I think some quality time thinking about the code will find the problem area.

**Comment by contra** [comment]
Any ETA on fixing this? We aren't able to release gulp 4.0 until this is fixed

**Comment by iamstarkov** [comment]
im waitin gulp4 for almost one year, pls fix it 🐱

**Comment by iarna** [comment]
I would love the gulp folks to provide some explanation for why this is a blocker for them. This was, if anything, _worse_ in npm@2.

**Comment by iarna** [comment]
Looking back in https://github.com/gulpjs/gulp-cli/issues/7 it looks like npm@2 used to just explode. With #8996 this may be _good enough_. 

And digging into https://github.com/gulpjs/gulp-cli/issues/7 I'd really like to see this reproduced with a current release, as I'm skeptical that it can still happen– binaries are linked into place by the `build` action, and that's done serially (and our serial actions happen in a deterministic order):
```
    [doSerialActions, 'build', staging, todo, trackLifecycle.newGroup('build')],
```

**Comment by heikki** [comment]
```sh
#!/bin/sh

counter=0

function install {
	let counter++
	# install until it fails
	npm install gulp gulp-cli > /dev/null && printf "install $counter, gulp bin -> " && readlink node_modules/.bin/gulp && install
}

echo "Running 'npm install gulp gulp-cli' repeatedly"
rm -fr node_modules
install
```

With npm 3.2.2 I get these results:

```sh
Running 'npm install gulp gulp-cli' repeatedly
install 1, gulp bin -> ../gulp-cli/bin/gulp.js
install 2, gulp bin -> ../gulp/bin/gulp.js
install 3, gulp bin -> ../gulp/bin/gulp.js
install 4, gulp bin -> ../gulp/bin/gulp.js
install 5, gulp bin -> ../gulp/bin/gulp.js
install 6, gulp bin -> ../gulp/bin/gulp.js
install 7, gulp bin -> ../gulp/bin/gulp.js
```

**Comment by iarna** [comment]
Ok, what I see is deterministic ordering for the initial run, and for the additional runs, but a difference between the first run and those additional runs.

**Comment by iarna** [comment]
So followup runs consistently produce this order:
```
npm sill diffTrees update gulp-cli@0.3.0
npm sill diffTrees update gulp@3.9.0
```

**Comment by iarna** [comment]
Likewise a fresh install consistently produces this order: (gulp and gulp-cli are installed last after allll their deps)
```
npm sill diffTrees add gulp@3.9.0
npm sill diffTrees add gulp-cli@0.3.0
``

**Comment by iarna** [comment]
I think the latter is happening due to gulp being a dev-dep of gulp-cli. The question is, why doesn't this rule also apply with the first scenario.

**Comment by iarna** [comment]
Scheduling this for Aug 20th's release.

**Comment by phated** [comment]
Excellent, thanks!

**Comment by othiym23** [comment]
Landed as d395a6b8767853eede6ad4502266ac99504f51a8, included in `npm@3.3.1`. Thanks for reporting this and for your patience, Blaine!

**Comment by phated** [comment]
:fireworks: :cake: @othiym23 @iarna
